### ISSUE TITLE
any reason why task**json needs "RepetitionTime" key?

### ISSUE THREAD

**Issue by satra** [poldracklab/fitlins#223]
There was an analysis being run, where it crashed because the task json file did not have RepetitionTime. I thought this would only be relevant to the nifti file json not the task json.

it ran fine after adding RepetitionTime to the task json files. doing this seems to increase failure surface.

**Comment by effigies** [comment]
Because JSON is more accessible and easier to fix, we've used RepetitionTime as the canonical source.

**Comment by satra** [comment]
but this is the task json, the timeseries json should still have RepetitionTime no, and isn't that the more definitive source?

**Comment by effigies** [comment]
Oh, if that's the case, that sounds like a pybids bug, failing either to collect the metadata or apply the inheritance principle. We should be pulling the metadata associated with the NIFTI file. Could you give a small tree that we could simulate to reproduce?

**Comment by satra** [comment]
i also found at that an earlier version of fitlins is being used. let me check if this error still happens with the current version.

**Comment by adelavega** [comment]
That's definitely a pybids issue. fitlins uses the `Analysis.get_design_matrix` entities that are provided. Pybids should be using the inheritence principle to get the correct meta-data. 

Definitely worth checking the fitlins/pybids version though as things have changed.

**Comment by satra** [comment]
@effigies and @adelavega - seeing this error in both fitlins master and latest on dockerhub. (@effigies - does fitlins have a drop into debugger mode, or change the nipype crash format via a config so i can check what `img_f` is?)

```
    sys.exit(run_fitlins(sys.argv[1:]))
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/fitlins/cli/run.py", line 263, in run_fitlins
    analysis.setup(**selectors)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/analysis/analysis.py", line 94, in setup
    b.setup(input_nodes, drop_na=drop_na, **selectors)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/analysis/analysis.py", line 207, in setup
    **kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/layout/layout.py", line 1114, in get_collections
    skip_empty=skip_empty, **kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/variables/io.py", line 93, in load_variables
    dataset = _load_time_variables(layout, dataset, scope=scope, **_kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/variables/io.py", line 177, in _load_time_variables
    tr = layout.get_metadata(img_f, scope=scope)['RepetitionTime']
KeyError: 'RepetitionTime'
```

versions:

```
Singularity> /neurodocker/startup.sh python -c "import bids; print(bids.__version__)"
0.10.2
Singularity> /neurodocker/startup.sh fitlins --version
fitlins v0.6.2+15.g64f124f
```

command line:

```
Singularity> /neurodocker/startup.sh fitlins -m /data/models/gambling.json -w /working \
--smoothing 4:l1:iso /data /working/out participant --participant-label inside7002 \
-d /data/derivatives/fmriprep --n-cpus 10 --mem-gb 65 --database-path /working/db \
-vvv --debug
```

<details>
<summary> full crashfile </summary>

```
Node: fitlins_wf.loader
Working directory: /working/fitlins_wf/loader

Node inputs:

database_path = /working/db
model = {'Name': 'gambling', 'Description': 'Model for gambling task', 'Input': {'task': 'gambl
ing'}, 'Steps': [{'Level': 'run', 'Transformations': [{'Name': 'Factor', 'Input': ['trial_type'
]}, {'Name': 'Or', 'Input': ['trial_type.loss', 'trial_type.reward'], 'output': ['all_blocks']}, {'Name': 'Convolve', 'Input': ['all_blocks']}], 'Model': {'X': ['all_blocks', 'framewise_disp
lacement', 'trans_x', 'trans_y', 'trans_z', 'rot_x', 'rot_y', 'rot_z']}, 'Contrasts': [{'Name':
 'all_vs_baseline', 'ConditionList': ['all_blocks'], 'weights': [1], 'type': 't'}]}, {'Level': 'subject', 'Contrasts': {'Conditions': ['all_vs_baseline'], 'Type': 't'}}, {'Level': 'dataset',
 'Contrasts': {'Conditions': ['all_vs_baseline'], 'Type': 't'}}]}
selectors = {'desc': 'preproc', 'space': 'MNI152NLin2009cAsym', 'subject': ['inside7002']}

Traceback (most recent call last):
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/plugins/mu
ltiproc.py", line 148, in __init__
    mp_context=mp_context,
TypeError: __init__() got an unexpected keyword argument 'initializer'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/plugins/mu
ltiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nod
es.py", line 516, in run
    result = self._run_interface(execute=True)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nod
es.py", line 635, in _run_interface
    return self._run_command(execute)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nod
es.py", line 741, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/interfaces/base/cor
e.py", line 397, in run
    runtime = self._run_interface(runtime)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/fitlins/interfaces/bids.py
", line 220, in _run_interface
    analysis.setup(drop_na=False, **selectors)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/analysis/analysis.py"
, line 94, in setup
    b.setup(input_nodes, drop_na=drop_na, **selectors)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/analysis/analysis.py"
, line 207, in setup
    **kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/layout/layout.py", li
ne 1114, in get_collections
    skip_empty=skip_empty, **kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/variables/io.py", lin
e 93, in load_variables
    dataset = _load_time_variables(layout, dataset, scope=scope, **_kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/variables/io.py", lin
e 177, in _load_time_variables
    tr = layout.get_metadata(img_f, scope=scope)['RepetitionTime']
KeyError: 'RepetitionTime'
```
</details>

**Comment by effigies** [comment]
I have not made crashfile format easily toggleable. IIRC the pklz were previously considered useless, so it's hard coded:

https://github.com/poldracklab/fitlins/blob/2d2fc554c3a58ad2d9a662798427ab6ebc6e6a81/fitlins/cli/run.py#L240

Should probably switch to a ranked precedence:

1) User nipype config
2) Fitlins-default nipype config
3) Nipype's default config

----

Does nipype have a mode where failures drop into a debugger? If so, I'm glad to enable it, but I don't recall ever running across it. Probably wouldn't be too hard to do: on crash, re-add node to queue as `run_without_submitting` and with `pdb` enabled.

**Comment by satra** [comment]
> Does nipype have a mode where failures drop into a debugger?

nipype cli has a mode for using pklz to drop into pdb when rerun. so pklz are not completely useless. they were they precisely for the scenario of debugging a node without having to rerun the whole workflow to drop into a debugger. we should open a different issue for this though. 

perhaps in this issue we can find out why a problem that seemed resolved in pybids is reappearing.

**Comment by effigies** [comment]
Okay. I'm working on #229 that will make it easier to set config files. In the meantime, is there any reason you can't install in a conda environment? I don't think we currently have any non-Python dependencies, and this will allow you to use an editable install and more easily do debugging.

**Comment by satra** [comment]
@effigies - is fitlins completely dependent on just python libraries (i had not checked)? if so, i can absolutely debug  via conda.

**Comment by effigies** [comment]
Yes. Masking used to be done with FSL, but it's not right now.

**Comment by tyarkoni** [comment]
@satra `img_f` is supposed to be a nifti image filename, I believe. My guess is that for some reason the wrong file (perhaps non-nifti) is being passed, so when it goes to get the metadata, it's reading the wrong sidecar. It's unclear to me from the above where the issue is introduced though.

I'm also wondering if we should prioritize merging https://github.com/bids-standard/pybids/pull/548, because there have been some pretty big changes to the way the `Analysis` class works, and I don't want to spend time debugging something that might no longer be salient. I think the main blocker on that PR right now is making sure fitlins can work with the changes.

**Comment by satra** [comment]
so i can verify that after fixing the nested derivatives directories and a few more keys in the model, i was able to run it on two subjects. i will now test scaling up to 100+.

i do think it would be useful to generate a better error message than the KeyError one.  i'll leave this to @effigies as to whether that should be a new issue or we update this issue.

**Comment by adelavega** [comment]
I think that's probably a pybids issue to throw a nicer error

**Comment by tyarkoni** [comment]
It's not clear to me that this could be handled well in pybids without suffering scope creep... I see it more as a validator issue. If you structure a project in a non-BIDS-compliant way, pybids will behave unpredictably, but ideally running the data through the validator would catch that, as it's illegal for a nifti file not to have a JSON sidecar containing `RepetitionTime`.

**Comment by effigies** [comment]
It could be a scope creep, but fetching the full metadata dictionary and retrieving a single value is going to be very common. Given that the dictionary can't know the name of the file argument that generated it, it leaves each caller in the position of identifying the potential failure and promoting the exception to one that lets you know the file along with the missing term.

We could take a page from `ConfigParser` and pass enough information to get a single value:

```Python
>>> layout.get_metadata(img_f, scope=scope)
{...}
>>> layout.get_metadata(img_f, term="RepetitionTime", scope=scope)
KeywordError("No 'RepetitionTime' metadata associated with the file <path>")
```

(I'm assuming `scope` is kw-only, but it's not very important here.)

Another way to do this would be to return a subclass of dictionary:

```Python
class BIDSMetadata(dict):
    _source_file = None
    def __getitem__(self, key):
        try:
            return super().__getitem__(self, key)
        except KeyError as e:
            if self._source_file is not None:
                raise KeyError(f"No '{key}' metadata associated with the file '{self._source_file}'")
            raise e
```

**Comment by tyarkoni** [comment]
Oh, I see. Yeah, I'd be totally fine with appending the filename that generated the error; that's completely generic. I was just worrying about handling this as a special case (i.e., trying to explain to the user that it's mandatory for every nifti to have a JSON containing `RepetitionTime`, etc.).

**Comment by effigies** [comment]
I'm happy to implement either of these proposals in pybids.

@tyarkoni @adelavega Do either of you have a preference? The latter feels hackier, but it's also a free improvement for all downstream tools using `get_metadata()`.

**Comment by adelavega** [comment]
I think the 2nd option is acceptably hacky to me.

**Comment by effigies** [comment]
Closed (I think) in bids-standard/pybids#593. Please reopen if still an issue.
